druntime/src/core/internal/array/capacity.d

```diff
@@ -146,64 +146,194 @@ Lcontinue:
     return curCapacity / size;
 }
 
-// HACK: `nothrow` and `pure` is faked.
-private extern (C) void[] _d_arraysetlengthT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
-private extern (C) void[] _d_arraysetlengthiT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
+/**
+Resize a dynamic array by setting its `.length` property.
+
+Newly created elements are initialized based on their default value.
+If the array's elements initialize to `0`, memory is zeroed out. Otherwise, elements are explicitly initialized.
 
-/*
- * This template is needed because there need to be a `_d_arraysetlengthTTrace!Tarr` instance for every
- * `_d_arraysetlengthT!Tarr`. By wrapping both of these functions inside of this template we force the
- * compiler to create a instance of both function for every type that is used.
- */
+This function handles memory allocation, expansion, and initialization while maintaining array integrity.
 
-/// Implementation of `_d_arraysetlengthT` and `_d_arraysetlengthTTrace`
-template _d_arraysetlengthTImpl(Tarr : T[], T)
+---
+void main()
 {
-    private enum errorMessage = "Cannot resize arrays if compiling without support for runtime type information!";
-
-    /**
-     * Resize dynamic array
-     * Params:
-     *  arr = the array that will be resized, taken as a reference
-     *  newlength = new length of array
-     * Returns:
-     *  The new length of the array
-     * Bugs:
-     *   The safety level of this function is faked. It shows itself as `@trusted pure nothrow` to not break existing code.
-     */
-    size_t _d_arraysetlengthT(return scope ref Tarr arr, size_t newlength) @trusted pure nothrow
-    {
-        version (DigitalMars) pragma(inline, false);
-        version (D_TypeInfo)
+    int[] a = [1, 2];
+    a.length = 3; // Gets lowered to `_d_arraysetlengthT!(int)(a, 3, false)`
+}
+---
+
+Params:
+    arr         = The array to resize.
+    newlength   = The new value for the array's `.length`.
+    isShared    = Whether the array is shared (affects GC expansion logic).
+
+Returns:
+    The resized array with updated length and properly initialized elements.
+
+Throws:
+    OutOfMemoryError if allocation fails.
+*/
+
+size_t _d_arraysetlengthT(Tarr : T[], T)(return ref scope Tarr arr, size_t newlength, bool isShared) @trusted
```

Comment:
```suggestion
size_t _d_arraysetlengthT(Tarr : T[], T)(return ref scope Tarr arr, size_t newlength) @trusted
```
According to this discussion [1], the preferred approach is to send the qualified type to DRuntime and create a small template wrapper whose purpose is to `Unqual` the type, then forward the result to the implementation. So roll back your compiler changes and pass the qualified array to the hook.

[1] https://github.com/dlang/dmd/pull/21205/#discussion_r2046733404